src/uapi.c

```diff
@@ -0,0 +1,323 @@
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ * Written by Alberto Planas <aplanas@suse.com>, Olaf Kirch <okir@suse.com>
+ */
+
+
+#include <stdlib.h>
+#include <sys/dir.h>
+#include <sys/param.h>
+#include <sys/utsname.h>
+#include <limits.h>
+#include <assert.h>
+#include <ctype.h>
+#include <errno.h>
+
+#include "uapi.h"
+#include "util.h"
+
+#define UAPI_LINE_MAX		1024
+
+
+static int	vercmp(const void *va, const void *vb);
+
+static uapi_boot_entry_t *
+uapi_boot_entry_new(void)
+{
+	uapi_boot_entry_t *ube;
+
+	ube = calloc(1, sizeof(*ube));
+	return ube;
+}
+
+static void
+drop_boot_entry(uapi_boot_entry_t **entry_p)
+{
+	if (*entry_p) {
+		uapi_boot_entry_free(*entry_p);
+		*entry_p = NULL;
+	}
+}
+
+static uapi_boot_entry_t *
+uapi_boot_entry_load(const char *path)
+{
+	uapi_boot_entry_t *result = NULL;
+	char line[UAPI_LINE_MAX];
+	FILE *fp;
+
+	if (!(fp = fopen(path, "r"))) {
+		error("Unable to open %s: %m\n", path);
+		return NULL;
+	}
+
+	result = uapi_boot_entry_new();
+	while (fgets(line, sizeof(line), fp)) {
+		char *key, *value;
+		unsigned int i;
+
+		if (!isalpha(line[0]))
+			continue;
+
+		/* strip white space off the end */
+		i = strlen(line);
+		while (i > 0 && isspace(line[i-1]))
+			line[--i] = '\0';
+
+		key = line;
+		for (i = 0; line[i]; ++i) {
+			if (isspace(line[i])) {
+				line[i++] = '\0';
+				break;
+			}
+		}
+
+		while (isspace(line[i]))
+			++i;
+		value = line + i;
+		if (*value == '\0')
+			value = NULL;
+
+                if (!strcmp("sort-key", key))
+			assign_string(&result->sort_key, value);
+                else
+                if (!strcmp("machine-id", key))
+			assign_string(&result->machine_id, value);
+                else
+                if (!strcmp("version", key))
+			assign_string(&result->version, value);
+                else
+                if (!strcmp("options", key))
+			assign_string(&result->options, value);
+                else
+                if (!strcmp("linux", key))
+			assign_string(&result->image_path, value);
+                else
+                if (!strcmp("initrd", key))
+			assign_string(&result->initrd_path, value);
+        }
+
+        fclose(fp);
+	return result;
+}
+
+static bool
+uapi_boot_entry_applies(const uapi_boot_entry_t *entry, const char *machine_id, const char *architecture)
+{
+	if (entry->machine_id && machine_id && strcmp(entry->machine_id, machine_id))
+		return false;
+	if (entry->architecture && architecture && strcmp(entry->architecture, architecture))
+		return false;
+	/* We don't check the efi key because it can only express "entry is EFI only"; it cannot express
+	 * "this is a legacy BIOS entry" */
```

Comment:
I think that this comment can be a bit confusing. The `efi` key can be indicated instead of `linux`, to execute a different EFI binary. AFAIK it does not indicate anything about the system (sd-boot only works on EFI, as it is itself another EFI binary)

Reply:
You're right, I misinterpreted the description in the uapi spec.